thumbor/engines/__init__.py

```diff
@@ -231,29 +232,32 @@ def get_proportional_height(self, new_width):
         width, height = self.size
         return round(float(new_width) * height / width, 0)
 
-    def _get_exif_segment(self):
+    def _get_exif_dict(self):
+        exif_object = self._get_exif_object()
+        if exif_object is None:
+            return
+
+        return exif_object._data
+
+    def _get_exif_object(self):
         if (not hasattr(self, 'exif')) or self.exif is None:
             return None
 
-        try:
-            exif_dict = piexif.load(self.exif)
-        except Exception:
-            logger.exception('Ignored error handling exif for reorientation')
-        else:
-            return exif_dict
-        return None
+        im = Image.new('RGB', (1, 1))
+        save_buffer = BytesIO()
+        im.save(save_buffer, format="JPEG", exif=self.exif)
+        return Image.open(save_buffer).getexif()
```

Comment:
The code above is bit expensive but it does work to get us an editable `PIL.Image.Exif` object. It's also portable so other engines can use it, like the opencv-engine (pillow was already a dependency). LMK if you can think of a better way to do this.

Reply:
Making a fake image with exif data seems like overkill. You could just use Exif class directly.
Also not sure why you want to access to private `_data` field. That looks fragile. I think we can use Exif object directly.

```diff
Index: thumbor/engines/__init__.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- thumbor/engines/__init__.py	(date 1562192633000)
+++ thumbor/engines/__init__.py	(date 1562196503403)
@@ -9,8 +9,7 @@
 # Copyright (c) 2011 globo.com thumbor@googlegroups.com
 
 from xml.etree.ElementTree import ParseError
-from io import BytesIO
-from PIL import Image
+from PIL.Image import Exif
 
 try:
     import cairosvg
@@ -232,21 +231,12 @@
         width, height = self.size
         return round(float(new_width) * height / width, 0)
 
-    def _get_exif_dict(self):
-        exif_object = self._get_exif_object()
-        if exif_object is None:
-            return
-
-        return exif_object._data
-
     def _get_exif_object(self):
-        if (not hasattr(self, 'exif')) or self.exif is None:
-            return None
+        exif = Exif()
+        if (hasattr(self, 'exif')) and self.exif is not None:
+            exif.load(self.exif)
 
-        im = Image.new('RGB', (1, 1))
-        save_buffer = BytesIO()
-        im.save(save_buffer, format="JPEG", exif=self.exif)
-        return Image.open(save_buffer).getexif()
+        return exif
 
     def get_orientation(self):
         """
@@ -255,10 +245,8 @@
         :return: Orientation value (1 - 8)
         :rtype: int or None
         """
-        exif_dict = self._get_exif_dict()
-        if exif_dict and exif_dict.get(274):
-            return exif_dict[274]
-        return None
+        exif = self._get_exif_object()
+        return exif.get(0x0112) or None
 
     def reorientate(self, override_exif=True):
         """
@@ -295,7 +283,7 @@
         if orientation != 1 and override_exif:
             exif = self._get_exif_object()
             exif[0x0112] = 1
-            self.exif = exif
+            self.exif = exif.tobytes()
 
     def gen_image(self, size, color):
         raise NotImplementedError()
Index: tests/engines/test_base_engine.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- tests/engines/test_base_engine.py	(date 1562192633000)
+++ tests/engines/test_base_engine.py	(date 1562196346827)
@@ -197,7 +197,7 @@
         expect(self.engine.get_orientation()).to_be_null()
 
     def test_get_orientation_without_orientation_in_exif(self):
-        self.engine.exif = 'Exif\x00\x00II*\x00\x08\x00\x00\x00\x04\x00\x1a\x01\x05\x00\x01\x00\x00\x006\x00\x00\x00\x1b\x01\x05\x00\x01\x00\x00\x00>\x00\x00\x00(\x01\x03\x00\x01\x00\x00\x00\x02\x00\x00\x00\x13\x02\x03\x00\x01\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x01\x00\x00\x00H\x00\x00\x00\x01\x00\x00\x00'
+        self.engine.exif = b'Exif\x00\x00II*\x00\x08\x00\x00\x00\x04\x00\x1a\x01\x05\x00\x01\x00\x00\x006\x00\x00\x00\x1b\x01\x05\x00\x01\x00\x00\x00>\x00\x00\x00(\x01\x03\x00\x01\x00\x00\x00\x02\x00\x00\x00\x13\x02\x03\x00\x01\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x01\x00\x00\x00H\x00\x00\x00\x01\x00\x00\x00'
         expect(self.engine.get_orientation()).to_be_null()
 
     def test_get_orientation(self):
```

Reply:
That's a great idea. Though I'm not sure why these tests are failing https://travis-ci.org/thumbor/thumbor/jobs/557429663#L1506

Reply:
let me take a look. 
It was fine running tests on my local before I submitted suggestion.

Reply:
@kkopachev Any luck?

Reply:
Not really a luck, but I figured it's Pillow's issue. Even with your original code it would fail.
Only reason it does not fail, because you put `Exif` object (not exif bytes) back into `self.exif` property. And it is luck that it's not used when saving image down the line, because `PRESERVE_EXIF_INFO` is off for these tests.
I'll dig around a bit more and probably going to open a ticket in Pillow 
:cry:

Reply:
simple repro script
```python
from PIL import Image, ImageOps

im = Image.open('tests/fixtures/images/10_years_of_Wikipedia_by_Guillaume_Paumier.jpg')
im = ImageOps.exif_transpose(im)
```